internal/instance/resource_instance.go

```diff
@@ -371,13 +384,73 @@ func (r InstanceResource) ValidateConfig(ctx context.Context, req resource.Valid
 		)
 	}
 
-	if !config.Image.IsNull() && !config.SourceInstance.IsNull() {
+	if !atMostOne(!config.Image.IsNull(), !config.SourceFile.IsNull(), !config.SourceInstance.IsNull()) {
```

Comment:
I consider this harder to read than before, could we make this more readable?

Reply:
The reason for this is, that we now have 3 different options, from which at most one (0 or 1) may be used.

I see the following ways to change this code:

1.
```suggestion
	if moreThanOne(!config.Image.IsNull(), !config.SourceFile.IsNull(), !config.SourceInstance.IsNull()) {
```
This removes one negation `!`, which might make it a little bit more readable. On the other side, it is pretty common in Go to have negative condition to enter an error path (basically stating the condition positively and then negating it for error path).

2.
```suggestion
	if maxOneEnabled(config.Image.IsNull(), config.SourceFile.IsNull(), config.SourceInstance.IsNull()) {
```
This removes all the negations. The name could also be `maxOneNonNull`.

3.
```suggestion
	if (!config.Image.IsNull() && (!config.SourceFile.IsNull() || !config.SourceInstance.IsNull())) || (!config.SourceFile.IsNull() && !config.SourceInstance.IsNull()) {
```
As list of boolean conditions. Might work for now, but if we need to extend it to a fourth case, I expect this solution to become pretty unreadable.

Which of those do you prefer?

Reply:
The second and in other programming languages this is method is usually called `any(…)`. Where it returns true if one of the values passed are true, otherwise false.

Reply:
After some more thinking, I hope I have now found a solution, that suits you in regards to readability.

Instead of having the `IsNull` part in the condition, it is now moved to the helper function. And I still have the initial `!`, which does line up well with the pattern in go, that the error path is triggered, if a condition is not met (e.g. `err != nil` or `!ok`).

The challenge is, that we need a function, that allows exactly zero or one of the config blocks to be present (enabled) and I would like to have a function, which also scales to more additions (currently we check 3 mutual exclusive config blocks, but in the future it could be even more).

`any` does not do the trick in my opinion, since `any` is expected to return true also if more than one element meets the criteria. Additionally, `any` is off the table for Go, since it is already an alias for `interface{}` and I don't assume, that we would like to export this function.